subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publisher/IvyDescriptorFileGenerator.java

```diff
@@ -263,6 +263,10 @@ private void writeDependencies(OptionalAttributeXmlWriter xmlWriter) throws IOEx
                     .attribute("rev", resolvedVersion != null ? resolvedVersion : dependency.getRevision())
                     .attribute("conf", dependency.getConfMapping());
 
+            if(resolvedVersion != null) {
+                xmlWriter.attribute("revConstraint", dependency.getRevision());
```

Comment:
If I understand correctly what's [documented in Ivy](http://ant.apache.org/ivy/history/2.1.0/ivyfile/dependency.html), `revConstraint` really makes sense if the `dependency.getRevision()` represents a dynamic selector.
However, even for a fixed version in Gradle, you could have the resolved version be different due to conflict resolution with transitive dependencies.

So I think the change cannot be that simple, there needs to be a confirmation that the declared version is indeed dynamic before putting it in `revConstraint`.

And to be complete, I wonder what would happen in Ivy if the resolved version was out of the bounds of the range, which is a valid resolution result with Gradle, unless `strictly` is used.

Reply:
Hi @ljacomet , good catch. I got biased by our use case and totally forgot about fixed versions.

Hmmm... I was looking at the `IvyDescriptorFileGenerator` and can't find a way to determine if a dependency was dynamic or not. `DefaultIvyDependency` doesn't seem to have this kind of information without computing it. Is there a util class that could be use here? or do you have any suggestion for this? I guess it would also make sense to write the revConstraint if resolved version is different from revision, something like:

```
if(resolvedVersion != null && resolvedVersion != dependency.getRevision()) {
                xmlWriter.attribute("revConstraint", dependency.getRevision());
            }
```

This would solve the first part, only write the `revConstraint` if `resolvedVersion` is different (in this case assumption is resolved from a dynamic/range)

for the 

```
what would happen in Ivy if the resolved version was out of the bounds of the range, which is a valid resolution result with Gradle, unless strictly is used.
```

I'm not sure of the expected behavior there. However, I think that the ivy descriptor file generator shouldn't be responsible for detemrining this information, wonder if this work should be done way before and just have the file generator to write.

Reply:
I just did the change I mentioned in https://github.com/gradle/gradle/pull/9121/commits/185f673c6aa7b8e7ff568f4533f86e1cde1444bc

Reply:
For determining if a version is dynamic, you could use `VersionRangeSelector.ALL_RANGE` similarly as in https://github.com/gradle/gradle/blob/bf9ab8ad1f0818311faeb57bddd983dc2217ddea/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/strategy/DefaultVersionSelectorScheme.java#L38

Reply:
thanks for the feedback @ljacomet , you can see the changes and new test case in https://github.com/gradle/gradle/pull/9121/commits/a89791bfbdc43c04c9b7228aad5971c936fac028